balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,286 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package balanceraggregator implements a BalancerAggregator helper.
+//
+// # Experimental
+//
+// Notice: This package is EXPERIMENTAL and may be changed or removed in a
+// later release.
+package balanceraggregator
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/balancer/gracefulswitch"
+	"google.golang.org/grpc/internal/grpcrand"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+// ChildState is the balancer state of a child along with the endpoint which
+// identifies the child balancer.
+type ChildState struct {
+	Endpoint resolver.Endpoint
+	State    balancer.State
+}
+
+// NewBalancer returns a new BalancerAggregator.
+func NewBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return &BalancerAggregator{
+		cc:       cc,
+		bOpts:    opts,
+		children: resolver.NewEndpointMap(),
+	}
+}
+
+// BalancerAggregator is a balancer that wraps child balancers. It creates a
+// child balancer with child config for every unique Endpoint received. It
+// updates the child states on any update from parent or child.
+type BalancerAggregator struct {
+	cc    balancer.ClientConn
+	bOpts balancer.BuildOptions
+
+	children *resolver.EndpointMap
+
+	// inhibitChildUpdates is set during UpdateClientConnState/ResolverError
+	// calls (calls to children will each produce an update, only want one
+	// update).
+	inhibitChildUpdates bool
```

Comment:
This also needs to be protected by a mutex, right?  It's set by calls coming in from the parent and read by concurrent updates from children.

Reply:
Or actually it's probably fine if it's atomic instead.

The reason we want to quell the updates from children during the update is that it's churn.  Having an async call come in from a child that has an update come out during the UpdateCCS call should be fine, unless it's hard to make the parent LB policy able to handle that (but I doubt that?).

Reply:
Yeah I'll make it atomic. The update during UpdateCCS call is part of my middle scenario I wrote in my reply to your synchronization comment. Thanks.